common/utils/include/claragenomics/utils/graph.hpp

```diff
@@ -0,0 +1,300 @@
+    virtual std::string get_node_label(node_id_t node) const
+    {
+        auto found_node = node_labels_.find(node);
+        if (found_node != node_labels_.end())
+        {
+            return found_node->second;
+        }
+        else
+        {
+            return "";
+        }
+    }
+
+    /// \brief Serialize graph structure to dot format
+    ///
+    /// \return A string encoding the graph in dot format
+    virtual std::string serialize_to_dot() const = 0;
+
+protected:
+    Graph() = default;
+
+    /// \brief Check if a directed edge exists in the grph
+    ///
+    /// \param edge A directed edge
+    /// \return Boolean result of check
+    bool directed_edge_exists(edge_t edge)
+    {
+        auto find_edge = edges_.find(edge);
+        if (find_edge == edges_.end())
+        {
+            return false;
+        }
+        else
+        {
+            return true;
+        }
+    }
+
+    /// \brief Update the adjacent nodes based on edge information
+    ///
+    /// \param edge A directed edge
+    void update_adject_nodes(edge_t edge)
+    {
+        auto find_node = adjacent_nodes_.find(edge.first);
+        if (find_node == adjacent_nodes_.end())
+        {
+            adjacent_nodes_.insert({edge.first, {edge.second}});
+        }
+        else
+        {
+            find_node->second.push_back(edge.second);
+        }
+    }
+
+    /// \brief Serialize node labels to dot format
+    ///
+    /// \param dot_str Output string stream to serialize labels to
+    void node_labels_to_dot(std::ostringstream& dot_str) const
+    {
+        const std::vector<node_id_t> nodes = get_node_ids();
+        for (auto node : nodes)
+        {
+            auto label_found = node_labels_.find(node);
+            if (label_found != node_labels_.end())
+            {
+                dot_str << node << " [label=\"" << label_found->second << "\"];\n";
+            }
+        }
+    }
+
+    /// \brief Serialize edges to dot format
+    ///
+    /// \param dot_str Output string stream to serialize labels to
+    /// \param node_separator DOT delimiter for edge description
+    void edges_to_dot(std::ostringstream& dot_str, const std::string& node_separator) const
+    {
+        for (auto iter : edges_)
+        {
+            const edge_t& edge          = iter.first;
+            const edge_weight_t& weight = iter.second;
+            dot_str << edge.first << " " << node_separator << " " << edge.second;
+            dot_str << " [label=\"" << weight << "\"];\n";
+        }
+    }
+
+    /// List of adjacent nodes per node ID
+    std::unordered_map<node_id_t, std::vector<node_id_t>> adjacent_nodes_;
+
+    /// All edges in the graph
+    std::unordered_map<edge_t, edge_weight_t, PairHash> edges_;
+
+    /// Label per node
+    std::unordered_map<node_id_t, std::string> node_labels_;
+
+    /// An empty list representing no connectivity
+    const std::vector<node_id_t> empty_;
```

Comment:
After I started making this change realized that to initialize the static member I would need to put it outside the header file, otherwise will get multiple definitions errors... Was hoping to keep this header only